File: address-geocode-canada-python/REST/get_geo_location_by_municipality_province_rest.py
import requests
from agca_response import GetGeoLocationByMunicipalityProvinceResponse, Error

# Endpoint URLs for ServiceObjects Address Geocode CA (AGCA) API
primary_url = "https://sws.serviceobjects.com/GCC/api.svc/json/GetGeoLocationByMunicipalityProvince?"
backup_url = "https://swsbackup.serviceobjects.com/GCC/api.svc/json/GetGeoLocationByMunicipalityProvince?"
trial_url = "https://trial.serviceobjects.com/GCC/api.svc/json/GetGeoLocationByMunicipalityProvince?"

def get_geo_location_by_municipality_province(
    municipality: str,
    province: str,
    license_key: str,
    is_live: bool = True
) -> GetGeoLocationByMunicipalityProvinceResponse:
    """
    Call ServiceObjects Address Geocode CA (AGCA) API's GetGeoLocationByMunicipalityProvince endpoint
    to retrieve geocoding information (latitude, longitude, postal code) for a given Canadian municipality and province.

    Parameters:
        municipality: The municipality of the address to geocode (e.g., "Cayuga").
        province: The province of the address to geocode (e.g., "ON").
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.

    Returns:
        GetGeoLocationByMunicipalityProvinceResponse: Parsed JSON response with geocoding results or error details.

    Raises:
        RuntimeError: If the API returns an error payload.
        equests.RequestException: On network/HTTP failures (trial mode).
    """
    params = {
        "Municipality": municipality,
        "Province": province,
        "LicenseKey": license_key,
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error', None)
        if not (error is None or error.get('Number', None) != "4"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"AGCA service error: {data['Error']}")

        # Convert JSON response to GetGeoLocationByMunicipalityProvinceResponse for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None

        return GetGeoLocationByMunicipalityProvinceResponse(
            Latitude=data.get("Latitude"),
            Longitude=data.get("Longitude"),
            PostalCode=data.get("PostalCode"),
            Error=error,
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"AGCA backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None

                return GetGeoLocationByMunicipalityProvinceResponse(
                    Latitude=data.get("Latitude"),
                    Longitude=data.get("Longitude"),
                    PostalCode=data.get("PostalCode"),
                    Error=error,
                )
            except Exception as backup_exc:
                raise RuntimeError("AGCA service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"AGCA trial error: {str(req_exc)}") from req_exc

File: address-geocode-canada-nodejs/REST/get_geo_location_by_municipality_province_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { GetGeoLocationByMunicipalityProvinceResponse } from './agca_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects Address Geocode CA (AGCA) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/GCC/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects Address Geocode CA (AGCA) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/GCC/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects Address Geocode CA (AGCA) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/GCC/api.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.Number is not equal to '4'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.Number !== '4';

/**
 * <summary>
 * Constructs a full URL for the GetGeoLocationByMunicipalityProvince API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}GetGeoLocationByMunicipalityProvince?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<GetGeoLocationByMunicipalityProvinceResponse>">A promise that resolves to a GetGeoLocationByMunicipalityProvinceResponse object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new GetGeoLocationByMunicipalityProvinceResponse(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects Address Geocode CA (AGCA) API's GetGeoLocationByMunicipalityProvince endpoint,
 * retrieving geocoding information (e.g., latitude, longitude, postal code) for a given Canadian municipality and province with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const GetGeoLocationByMunicipalityProvinceClient = {
    /**
     * <summary>
     * Asynchronously invokes the GetGeoLocationByMunicipalityProvince API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.Number == '4') in live mode.
     * </summary>
     * @param {Object} input - Input parameters for the API call.
     * @param {string} input.Municipality - The municipality to geocode (e.g., "Cayuga"). Required.
     * @param {string} input.Province - The province to geocode (e.g., "ON"). Required.
     * @param {string} input.LicenseKey - Your license key to use the service.
     * @param {boolean} input.IsLive - Value to determine whether to use the live or trial servers.
     * @param {number} input.TimeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<GetGeoLocationByMunicipalityProvinceResponse>} - A promise that resolves to a GetGeoLocationByMunicipalityProvinceResponse object.
     */
    async invokeAsync(input) {
        const params = {
            Municipality: input.Municipality,
            Province: input.Province,
            LicenseKey: input.LicenseKey
        };

        const url = buildUrl(params, input.IsLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, input.TimeoutSeconds);

        if (input.IsLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, input.TimeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the GetGeoLocationByMunicipalityProvince API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {Object} input - Input parameters for the API call.
     * @param {string} input.Municipality - The municipality to geocode (e.g., "Cayuga"). Required.
     * @param {string} input.Province - The province to geocode (e.g., "ON"). Required.
     * @param {string} input.LicenseKey - Your license key to use the service.
     * @param {boolean} input.IsLive - Value to determine whether to use the live or trial servers.
     * @param {number} input.TimeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {GetGeoLocationByMunicipalityProvinceResponse} - A GetGeoLocationByMunicipalityProvinceResponse object with geocoding details or an error.
     */
    invoke(Municipality, Province, LicenseKey, IsLive = true, TimeoutSeconds = 15) {
        return (async () => await this.invokeAsync({ Municipality, Province, LicenseKey, IsLive, TimeoutSeconds }))();
    }
};

export { GetGeoLocationByMunicipalityProvinceClient, GetGeoLocationByMunicipalityProvinceResponse };